fs-index/src/index.rs

```diff
@@ -1,1120 +1,501 @@
+        for entry in walker {
+            let entry = entry.map_err(|e| {
+                ArklibError::Path(format!("Error walking directory: {}", e))
+            })?;
+            // Ignore directories
+            if !entry.file_type().is_file() {
+                continue;
+            }
+            let path = entry.path();
+            let metadata = fs::metadata(path)?;
+            // Ignore empty files
+            if metadata.len() == 0 {
+                continue;
+            }
+            let last_modified = metadata.modified()?;
+            let id = Id::from_path(path)?;
+            // Path is relative to the root
+            let path = path.strip_prefix(&root).map_err(|_| {
+                ArklibError::Path("Error stripping prefix".to_string())
+            })?;
+
+            // Create the resource and add it to the index
+            let resource = IndexedResource {
+                id: id.clone(),
+                path: path.to_path_buf(),
+                last_modified,
+            };
+            path_to_resource.insert(resource.path.clone(), resource.clone());
+            id_to_resources
+                .entry(id)
+                .or_insert_with(Vec::new)
+                .push(resource);
+        }
 
-            assert_eq!(actual.root, path.clone());
-            assert_eq!(actual.path2id.len(), 0);
-            assert_eq!(actual.id2path.len(), 0);
-            assert_eq!(actual.collisions.len(), 0);
+        Ok(ResourceIndex {
+            root,
+            id_to_resources,
+            path_to_resource,
         })
     }
 
-    #[test]
-    fn should_not_index_1_empty_directory() {
-        run_test_and_clean_up(|path| {
-            create_dir_at(path.clone());
-
-            let actual: ResourceIndex<Crc32> =
-                ResourceIndex::build(path.clone());
+    /// Update the index with the latest information from the file system
+    pub fn update_all(&mut self) -> Result<IndexUpdate<Id>> {
+        log::debug!("Updating index at root path: {:?}", self.root);
+
+        let mut added = Vec::new();
+        let mut modified = Vec::new();
+        let mut removed = Vec::new();
+
+        let new_index = ResourceIndex::build(&self.root)?;
+
+        // Compare the new index with the old index
+        let current_resources = self.resources();
+        let new_resources = new_index.resources();
+        for resource in new_resources.clone() {
+            // If the resource is in the old index, check if it has been
+            // modified
+            if let Some(current_resource) =
+                self.get_resource_by_path(&resource.path)
+            {
+                if current_resource != &resource {
+                    modified.push(resource.clone());
+                }
+            }
+            // If the resource is not in the old index, it has been added
+            else {
+                added.push(resource.clone());
+            }
```

Comment:
The `resource.path` is not found in the index, but a resource with same content can be already tracked by another path. The app developer should treat additions only if the resource didn't exist before.

Probably we should implement "moved" event instead of "modified". Resources cannot be modified because they are identified by their content, if the content is changed then it's another resource. But resources can move, i.e. change their paths.

Reply:
> The `resource.path` is not found in the index, but a resource with same content can be already tracked by another path.

This suggests that we might want to use the file hash as the unique identifier. So far, we've been using paths to determine if it's the same file due to the file system making sure of no duplicate paths.

To use the resource ID as the sole unique identifier, we would need to prevent adding two files with the same content. Is this acceptable?

Reply:
The high-level goals are:
1. to find all different resources
2. determine their paths
3. detect if we have different paths for a single resource
4. list all different paths for same a single resource

1 and 2 are mandatory goals, 3 and 4 are optional goals.